Format the CI and CD stage spec (#53)
cdevents/spec: continuous-integration-pipeline-events.md

__Note:__ This is a work-in-progress draft version and is being worked on by members of the Events SIG. You are very welcome to join the work and the discussions!

Continuous Integration (CI) events include the subject and predicates related to CI activities such as building software, producing artifacts and running tests.

## Subjects

> This specification defines three subjects in this stage: `builds`, `artifacts` and `tests`. Events associated with these subjects are typically generated either by a CI system that orchestrates the process or by a specific build or test tool directly. Some artifact events may be triggered by the system that stores the artifact as well.

Reviewer: ```suggestion
This specification defines three subjects in this stage: `builds`, `artifacts` and `tests`. Events associated with these subjects are typically generated either by a CI system that orchestrates the process or by a specific build or test tool directly. Some artifact events may be generated by the system that stores the artifact as well.
```

Author: Good catch!

__Note:__ The data model for `testCase` only includes `id` and `source`, inputs and outputs of the process are not specified yet, as well as the relation to `testSuite`.

| Field | Type | Description | Examples |
|-------|------|-------------|----------|
| id    | `String` | Uniquely identifies the subject within the source. | `unitest-abc`, `e2e-test1`, `scan-image1` |
| source | `URI-Reference` | [source](../spec.md#source) from the context | `staging/tekton`, `tekton-dev-123`|

### `testSuite`

A `testSuite` represents a set of one or more `testCases`.

__Note:__ The data model for `testSuite` only includes `id` and `source`, inputs and outputs of the process are not specified yet, as well as the relation to `testCase`.

| Field | Type | Description | Examples |
|-------|------|-------------|----------|
| id    | `String` | Uniquely identifies the subject within the source. | `unit`, `e2e`, `security` |
| source | `URI-Reference` | [source](../spec.md#source) from the context | `staging/tekton`, `tekton-dev-123`|

### `artifact`

> An `artifact` is usually produced as output of a build pipeline. Events needs to be generated to indicate that an `artifact` have been packaged and released for others to use. These events can be produced by the artifact producer or by the artifact storage system.

Reviewer: ```suggestion
An `artifact` is usually produced as output of a build process. Events need to be generated to indicate that an `artifact` has been packaged and released for others to use. These events can be produced by the artifact producer or by the artifact storage system.
```

- Event Type: __`dev.cdevents.build.queued`__
- Predicate: queued
- Subject: [`build`](#build)

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `1234`, `maven123`, `builds/taskrun123` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

🚧 The build model is work in progress.

### `build started`

This event represents a Build task that has been started; this build process usually is in charge of producing a binary from source code.

- Event Type: __`dev.cdevents.build.started`__
- Predicate: started
- Subject: [`build`](#build)

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `1234`, `maven123`, `builds/taskrun123` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

🚧 The build model is work in progress.

### `build finished`

> This event represents a Build task that has finished; the event will contain the finished status, success, error or failure

Reviewer: So far there is no status/result field defined in the table below. Should it be omitted from this explanation, or mentioned as something not yet defined?

Author: I preserved the original text and data model, but yes, it's clearer like you proposed.

This event represents a Test task that has been queued, and it is waiting to be started.

- Event Type: __`dev.cdevents.testcase.queued`__
- Predicate: queued
- Subject: [`testCase`](#testcase)

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `unitest-abc`, `e2e-test1`, `scan-image1` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

🚧 The testCase model is work in progress.

### `testCase started`

This event represents a Test task that has started.

- Event Type: __`dev.cdevents.testcase.started`__
- Predicate: started
- Subject: [`testCase`](#testcase)

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `unitest-abc`, `e2e-test1`, `scan-image1` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

🚧 The testCase model is work in progress.

### `testCase finished`

> This event represents a Test task that has finished, the event will contain the finished status: success, error or failure.

Reviewer: Similar comment as to build finished, the table doesn't include any status/outcome field.

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `unit`, `e2e`, `security` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

🚧 The testSuite model is work in progress.

### `testSuite finished`

This event represents a Test suite that has has finished, the event will contain the finished status: success, error or failure.

- Event Type: __`dev.cdevents.testsuite.finished`__
- Predicate: finished
- Subject: [`testSuite`](#testsuite)

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `unit`, `e2e`, `security` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

🚧 The testSuite model is work in progress.

### `artifact packaged`

The event represents an artifact that has been packaged for distribution; this artifact is now versioned with a fixed version.

> 

Reviewer: WIP. We need to add a version identifier to this event. Or should it be part of the id string?

Author: I think this question links back to the question about the format of the artifact id: https://github.com/cdevents/spec/issues/38

Author: I would prefer to avoid adding any field in this PR.
In the PoCs so far we've included the version in the id.

- Event Type: __`dev.cdevents.artifact.packaged`__
- Predicate: packaged
- Subject: [`artifact`](#artifact)

| Field | Type | Description | Examples | Mandatory ✅ \| Optional ⚪ |
|-------|------|-------------|----------|----------------------------|
| id    | `String` | Uniquely identifies the subject within the source. | `abcde-12344-a2b35`, `sh0b31b1c02ff458ad9b7b81cbdf8f028bd54699fa151f221d1e8de6817db93427a1234`, `myimage@sha256:123456abcded`, `six-1.14.0-py2.py3-none-any.whl` | ✅ |
| source | `URI-Reference` | [source](../spec.md#source) from the context | | ⚪ |

### `artifact published`

The event represents an artifact that has been published and it can be advertised for others to use.

> 

Reviewer: WIP. We need to add a location to this event.

Author: Ditto - this goes back to the format of the artifact ID - but yeah, a url makes sense, I think we've used a url in the spec but stored it somewhere else since it wasn't in the spec.
